### PR TITLE
Feat/`TradeFee` and `BudgetChecker` documentation

### PR DESCRIPTION
**Pull request by mhrvth**
Documentation of fee accounting with the new `BudgetChecker` and the `TradeFee`

### CODE REVIEW

**Comment by mhrvth** [comment]
[sc-21346]

File: docs/developers/strategies/architecture.md [hunk 1]
```diff
@@ -159,3 +159,85 @@ Below are some of the user functions or properties under `OrderTracker` that you
...
+
+An order proposal created by a strategy.
+
+A `BudgetChecker` takes an `OrderCandidate` and fills in fees to be paid for this particular order (in pre-defined tokens). Then checks if, after accounting for the fees, the account balances allow for a placement of this order, and if not, adjusts the order amount accordingly. 
+
+Fees can be payable in the base tokens, quote tokens, or 3rd party tokens.
+
+If an order opens a position, fees will be charged as an additional cost of the order.
+
+If an order closes a position, fees will be deducted from the returns.
```
Comment (petioptrv) [inline docs/developers/strategies/architecture.md]: This part is generally, but not always, the case. For instance, Binance deducts the buy-order fees from the returns of a trade (e.g. if I create an order to buy 1,000 USDT worth of BTC, I get charged 1,000 USDT, but I will only get 999 USDT worth of BTC, with 1 USDT being the fee).

The user doesn't need to worry about this though, as each exchange can define their own fees schema and specify this behaviour. That being said, the docs should mention that the strategy has to get the budget checker from the exchange (i.e. `self.exchange.budget_checker`)

File: docs/developers/strategies/architecture.md [hunk 2]
```diff
@@ -159,3 +159,85 @@ Below are some of the user functions or properties under `OrderTracker` that you
...
+- `get_fee_impact_on_order_cost()`: returns order cost for a particular position opening `OrderCandidate` with fees accounted for
+- `get_fee_impact_on_order_returns()`: returns order returns for a particular position closing `OrderCandidate` with fees accounted for
+
+#### AddedToCostTradeFee
+
+Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`
+
+#### DeductedFromReturnsTradeFee
+
+Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`
```
Comment (petioptrv) [inline docs/developers/strategies/architecture.md]: This section is more relevant to the creation of exchanges, and strategies shouldn't worry about those details (except for the example below, which I think we should definitely keep). The info on `BudgetChecker` and `CandidateOrder` is all that's relevant to the strategies actually

File: docs/developers/strategies/architecture.md [hunk 3]
```diff
@@ -159,3 +159,85 @@ Below are some of the user functions or properties under `OrderTracker` that you
     Returns a dictionary of order IDs that are being cancelled.
 
     Return type: `Dict[str, float]`
+
+
+## Fee Accounting
+
+### OrderCandidate
```
Comment (petioptrv) [inline docs/developers/strategies/architecture.md:166]: On the strategy's use of the `OrderCandidate`, I think we can add something like, "Once the candidate order has been sized by the budget checker, the strategy can examine the sized order to get more information such as `OrderCandidate.collateral_dict` to get a dictionary of the costs associated with the order, or `OrderCandidate.potential_returns` to get an idea of the token and amount of the returns associated with the order."

**Review by petioptrv — CHANGES_REQUESTED** [review]

File: docs/developers/strategies/architecture.md (hunk 2 above)
Comment (mhrvth) [reply docs/developers/strategies/architecture.md]: Should this be removed entirely or moved to connectors?

Comment (petioptrv) [reply docs/developers/strategies/architecture.md]: Good point, moved to connectors would be the way to go, but then some more info should be added re the `TradeFeeSchema`, which is a way for connectors to define specific schemas (like Binance and the fee that's deducted from the returns of the buy orders). This also reminds that we should update the [configs override page](https://hummingbot.org/global-configs/override-fees/).

Comment (mhrvth) [reply docs/developers/strategies/architecture.md]: I've implemented the changes. Please check them out.

File: docs/developers/connectors/architecture.md [hunk 4]
```diff
@@ -146,3 +146,62 @@ For more details on how to begin implementing the components, please refer to th
...
+
+### TradeFee
+
+`TradeFee` is being used by the `BudgetChecker` to determine the appropriate amount and asset of fees to be applied to an order.
+An exchange needs to provide a `TradeFeeSchema` defining its fee structure. Based on this schema a `TradeFee` object will be constructed.
+
+#### TradeFeeSchema
+
+Contains the necessary information to build the `TradeFee` object.
+For both makers and takers specifies percent and fixed fees, and tokens in which the fees are paid.
```
Comment (petioptrv) [inline docs/developers/connectors/architecture.md]:
```suggestion
## Fee Accounting
The `BudgetChecker` uses the information from a `TradeFeeSchema` to generate a specific instance of `TradeFeeBase` that is then applied to an `OrderCandidate` in order to asses the order's effects on account balances.

### TradeFee

The `TradeFee` object contains the necessary information to account for fees when estimating an order's impact on account balances.

#### TradeFeeSchema

Contains the necessary information to build the `TradeFee` object.
For both makers and takers specifies percent and fixed fees, and tokens in which the fees are paid.
Exchanges must specify their respective default schemas inside their `[exchange]_utils.py` files:
```python
DEFAULT_FEES = TradeFeeSchema(
    maker_percent_fee_decimal=Decimal("0.001"),
    taker_percent_fee_decimal=Decimal("0.001"),
)
```
```

File: docs/developers/connectors/architecture.md [hunk 5]
```diff
@@ -146,3 +146,62 @@ For more details on how to begin implementing the components, please refer to th
...
+- `maker_percent_fee_decimal: Decimal`
+- `taker_percent_fee_decimal: Decimal`
+- `buy_percent_fee_deducted_from_returns: bool`
+- `maker_fixed_fees: List`
+- `taker_fixed_fees: List`
+
+
+#### TradeFeeBase
+
+The `TradeFeeBase` class uses a `TradeFeeSchema` object of the connector to determine fees to be applied - their types, amounts and assets.
```
Comment (petioptrv) [inline docs/developers/connectors/architecture.md]:
```suggestion
A specific instance of the `TradeFeeBase` class defines the fees to be applied to an order - their types, amounts and assets.
```

File: docs/developers/connectors/architecture.md [hunk 6]
```diff
@@ -146,3 +146,62 @@ For more details on how to begin implementing the components, please refer to th
...
+
+- `fee_amount_in_quote()`: calculates a total fee in quote asset units as a combination of a percentage fee and fixed fees
+- `get_fee_impact_on_order_cost()`: returns order cost for a particular position opening `OrderCandidate` with fees accounted for
+- `get_fee_impact_on_order_returns()`: returns order returns for a particular position closing `OrderCandidate` with fees accounted for
+
+#### AddedToCostTradeFee
+
+Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`
```
Comment (petioptrv) [inline docs/developers/connectors/architecture.md]:
```suggestion
Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`.
Fees of this class are applied on top of the cost of a buy order (e.g. a buy order of 10 COINX at 9 USDT with a fee of 1% means that the user's account will be deducted 90.9 USDT and added 10 COINX — this is most exchanges' approach to fees).

File: docs/developers/connectors/architecture.md [hunk 7]
```diff
@@ -146,3 +146,62 @@ For more details on how to begin implementing the components, please refer to th
...
+
+#### DeductedFromReturnsTradeFee
+
+Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`
```
Comment (petioptrv) [inline docs/developers/connectors/architecture.md]:
```suggestion
Extends `TradeFeeBase`, implements `get_fee_impact_on_order_cost()`, `get_fee_impact_on_order_returns()`.
Fees of this class are deducted from the returns of a buy order (e.g. a buy order of 10 COINX at 9 USDT with a fee of 1% means that the user's account will be deducted 90 USDT and added 9.9 COINX — this is Binance's approach to fees).

**Review by petioptrv — CHANGES_REQUESTED** [review]

File: docs/developers/connectors/architecture.md [hunk 8]
```diff
@@ -146,3 +146,70 @@ For more details on how to begin implementing the components, please refer to th
...
+#### TradeFeeSchema
+
+Contains the necessary information to build the `TradeFee` object.
+For both makers and takers specifies percent and fixed fees, and tokens in which the fees are paid.
+Exchanges must specify their respective default schemas inside their `[exchange]_utils.py` files:
+```python
+DEFAULT_FEES = TradeFeeSchema(
+    maker_percent_fee_decimal=Decimal("0.001"),
+    taker_percent_fee_decimal=Decimal("0.001"),
+)
```
Comment (petioptrv) [inline docs/developers/connectors/architecture.md]: The code snipped is not closed here, but it won't let me insert the closing ticks in a suggestion

**Review by petioptrv — CHANGES_REQUESTED** [review]

File: docs/developers/connectors/architecture.md (hunk 8 above)
Comment (mhrvth) [reply docs/developers/connectors/architecture.md]: Fixed

**Review by petioptrv — APPROVED** [review]
Changes LGTM!

**Review by dennisocana — APPROVED** [review]

**Comment by mhrvth** [comment]
Created a PR in the Foundation repo: https://github.com/hummingbot/hummingbot-site/pull/47

**Review by dennisocana — APPROVED** [review]

**Review by petioptrv — APPROVED** [review]

**Pull request merged**